lib/Doctrine/Annotations/Metadata/Builder/AnnotationMetadataBuilder.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace Doctrine\Annotations\Metadata\Builder;
+
+use Doctrine\Annotations\Metadata\AnnotationMetadata;
+use Doctrine\Annotations\Metadata\AnnotationTarget;
+use Doctrine\Annotations\Metadata\PropertyMetadata;
+
+/**
+ * @internal
+ */
+final class AnnotationMetadataBuilder
+{
+    /** @var string */
+    private $name;
+
+    /** @var AnnotationTarget */
+    private $target;
+
+    /** @var PropertyMetadata[] */
+    private $properties = [];
+
+    /** @var bool */
+    private $usesConstructor = false;
+
+    public function __construct(string $name)
+    {
+        $this->name   = $name;
+        $this->target = AnnotationTarget::all();
+    }
+
+    public function withTarget(AnnotationTarget $target) : self
+    {
+        $new         = clone $this;
```

Comment:
I get why you would want an immutable object for metadata, but looking at how the builder has to be used I find it cumbersome in this case. The DocParser is filled with repeated assignments to the same variable for something that is thrown away after calling `build()` on it. My suggestion would be to make the builder mutable but reset its state after calling `build()` to avoid issues.

Reply:
I agree mutable builder makes stuff simpler and more readable so I removed the immutability. Although it wasn't the source of performance degradation. :)